pkg/catalogv2/git/utils.go

```diff
@@ -0,0 +1,207 @@
+package git
+
+import (
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/hex"
+	"encoding/pem"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type basicRoundTripper struct {
+	username string
+	password string
+	next     http.RoundTripper
+}
+
+func (b *basicRoundTripper) RoundTrip(request *http.Request) (*http.Response, error) {
+	request.SetBasicAuth(b.username, b.password)
+	return b.next.RoundTrip(request)
+}
+
+func gitDir(namespace, name, gitURL string) string {
+	staticDir := filepath.Join(staticDir, namespace, name, hash(gitURL))
+	if s, err := os.Stat(staticDir); err == nil && s.IsDir() {
+		return staticDir
+	}
+	return filepath.Join(stateDir, namespace, name, hash(gitURL))
+}
+
+// isBundled check if the repositories are bundled on local static directory
+func isBundled(git *git) bool {
+	return strings.HasPrefix(git.Directory, staticDir)
+}
+
+// isGitSSH checks if the url is parsable to ssh url standard
+func isGitSSH(gitURL string) (bool, error) {
+	// Matches URLs with the format [anything]@[anything]:[anything]
+	return regexp.MatchString("(.+)@(.+):(.+)", gitURL)
+}
+
+func hash(gitURL string) string {
+	b := sha256.Sum256([]byte(gitURL))
+	return hex.EncodeToString(b[:])
+}
+
+// convertDERToPEM converts a src DER certificate into PEM with line breaks, header, and footer.
+func convertDERToPEM(src []byte) []byte {
+	return pem.EncodeToMemory(&pem.Block{
+		Type:    "CERTIFICATE",
+		Headers: map[string]string{},
+		Bytes:   src,
+	})
+}
+
+// formatGitURL is used by remoteSHAChange
+func formatGitURL(endpoint, branch string) string {
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return ""
+	}
+
+	pathParts := strings.Split(u.Path, "/")
+	switch u.Hostname() {
+	case "github.com":
+		if len(pathParts) >= 3 {
+			org := pathParts[1]
+			repo := strings.TrimSuffix(pathParts[2], ".git")
+			return fmt.Sprintf("https://api.github.com/repos/%s/%s/commits/%s", org, repo, branch)
+		}
+	case "git.rancher.io":
+		repo := strings.TrimSuffix(pathParts[1], ".git")
+		u.Path = fmt.Sprintf("/repos/%s/commits/%s", repo, branch)
+		return u.String()
+	}
+
+	return ""
+}
+
+// TODO: evaluate removal when possible,
+// gogit does not implement a similar method without fetching the repo what is a costly operation
```

Comment:
`go-git` does support `git ls-remote`, which does not require a full clone. Here's an example: https://github.com/fluxcd/pkg/blob/a3e50ffbcdb59c1e3e6bbd35d72b0337f23a74eb/git/gogit/clone.go#L411